File: leetcode438/leetcode438.py
from collections import Counter
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        ans = []
        cnt_p = Counter(p)
        cnt_s = Counter()
        for right, c in enumerate(s):
            cnt_s[c] += 1
            if right < len(p) - 1:
                continue
            if cnt_s == cnt_p:
                ans.append(right - len(p) + 1)
            cnt_s[s[right - len(p) + 1]] -= 1
        return ans

File: leetcode438/leetcode438.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
    public:
        vector<int> findAnagrams(string s, string p) {
            vector<int> ans;
            array<int, 26> cnt_p{};
            array<int, 26> cnt_s{};
            for (char c : p){
                cnt_p[c - 'a'] ++;
            }
            for (int right = 0; right < s.size(); right ++){
                cnt_s[s[right] - 'a'] ++;
                if (right < p.size() - 1){
                    continue;
                }
                if (cnt_s == cnt_p){
                    ans.push_back(right - p.size() + 1);
                }
                cnt_s[s[right - p.size() + 1] - 'a'] --;
            }
            return ans;
        }
    };
    
    